agent/xds/rbac.go

```diff
@@ -417,17 +462,47 @@ type rbacPermission struct {
 	ComputedPermission *envoy_rbac_v3.Permission
 }
 
+// Flatten ensure the permission rules, not-rules, and jwt validation rules are merged into a single computed permission.
+//
+// Details on JWTInfo section:
+// For each JWTInfo (AKA provider required), this builds 1 single permission that validates that the jwt has
+// the right issuer (`iss`) field and validates the claims (if any).
+//
+// After generating a single permission per info, it combines all the info permissions into a single OrPermission.
+// This orPermission is then attached to initial computed permission for jwt payload and claims validation.
 func (p *rbacPermission) Flatten() *envoy_rbac_v3.Permission {
-	if len(p.NotPerms) == 0 {
-		return p.Perm
+	computedPermission := p.Perm
+	if len(p.NotPerms) == 0 && len(p.jwtInfos) == 0 {
+		return computedPermission
 	}
 
-	parts := make([]*envoy_rbac_v3.Permission, 0, len(p.NotPerms)+1)
-	parts = append(parts, p.Perm)
-	for _, notPerm := range p.NotPerms {
-		parts = append(parts, notPermission(notPerm))
+	if len(p.NotPerms) != 0 {
+		parts := make([]*envoy_rbac_v3.Permission, 0, len(p.NotPerms)+1)
+		parts = append(parts, p.Perm)
+		for _, notPerm := range p.NotPerms {
+			parts = append(parts, notPermission(notPerm))
+		}
+		computedPermission = andPermissions(parts)
 	}
-	return andPermissions(parts)
+
+	if len(p.jwtInfos) == 0 {
+		return computedPermission
+	}
+
+	var jwtPerms []*envoy_rbac_v3.Permission
```

Comment:
I think this assembly routine is duplicated and could be refactored into a shared helper.